WNPRC_EHR/src/org/labkey/wnprc_ehr/notification/IrregularObsBehaviorNotification.java

```diff
@@ -0,0 +1,242 @@
+            long count = ts.getRowCount();
+            if (count > 0) {
+                msg.append("<p>There have been <strong>");
+                msg.append(count);
+                msg.append("</strong> behavior related irregular observations between ");
+                msg.append(startTime.format(displayDateFormat));
+                msg.append(" and ");
+                msg.append(endTime.format(displayDateFormat));
+                msg.append(".</p>");
+
+                TableInfo behaviorCodesTableInfo = QueryService.get().getUserSchema(u, c, "ehr_lookups").getTable("obs_behavior");
+
+                int maxBehaviorsLength = 0;
+                while (rs.next()) {
+                    String behaviors = getBehaviorString(behaviorCodesTableInfo, rs.getRowMap());
+                    if (behaviors != null && behaviors.length() > maxBehaviorsLength) {
+                        maxBehaviorsLength = behaviors.length();
+                    }
+                }
+                int behaviorsDashLength = (int) (maxBehaviorsLength * 1.75);
+
+                msg.append("<table>");
+
+                msg.append(createTableBorder(behaviorsDashLength));
+
+                msg.append("<tr>");
+                msg.append("<td>|</td>");
+                msg.append("<th>").append("Id").append("</th>");
+                msg.append("<td>|</td>");
+                msg.append("<th>").append("Room").append("</th>");
+                msg.append("<td>|</td>");
+                msg.append("<th>").append("Cage").append("</th>");
+                msg.append("<td>|</td>");
+                msg.append("<th>").append("Date/Time").append("</th>");
+                msg.append("<td>|</td>");
+                msg.append("<th>").append("Observation(s)").append("</th>");
+                msg.append("<td>|</td>");
+                msg.append("</tr>");
+
+                rs.beforeFirst();
+
+                while (rs.next()) {
+                    Map<String, Object> row = rs.getRowMap();
+                    String behaviors = getBehaviorString(behaviorCodesTableInfo, rs.getRowMap());
+
+                    StringBuilder hrefForAnimalAbstract = new StringBuilder();
+                    hrefForAnimalAbstract.append(new Path(ActionURL.getBaseServerURL(), "ehr", c.getPath(), "animalHistory.view")).toString();
+                    hrefForAnimalAbstract.append("?#subjects:").append(row.get("Id")).append("&inputType:singleSubject&showReport:1&activeReport:abstract");
+                    StringBuilder animalId = new StringBuilder();
+                    animalId.append("<a href='").append(hrefForAnimalAbstract).append("'>");
+                    animalId.append(row.get("Id"));
+                    animalId.append("</a>");
+
+                    msg.append(createTableBorder(behaviorsDashLength));
+
+                    msg.append("<tr>");
+                    msg.append("<td>|</td>");
+                    msg.append("<td>").append(animalId).append("</td>");
```

Comment:
Would this work:
HtmlString animalEncoded = HtmlString.of(animalId.toString());
...
...
msg.append("<td>|</td>");
 msg.append("<td>").append(animalEncoded).append("</td>");

Or should I encode other parts of the msg.

Reply:
You don't need to encode static non-HTML strings like "|", so that's safe.

There's no need to call msg.append(""), since adding the empty string will be a no-op. It's not an encoding or security concern, it's just useless code.

Reply:
Sorry Github remove the html encoding but here is what I was trying to proposed.

   
                    HtmlString animalEncoded = HtmlString.of((String)row.get("Id"));
                    StringBuilder hrefForAnimalAbstract = new StringBuilder();
                    hrefForAnimalAbstract.append(new Path(ActionURL.getBaseServerURL(), "ehr", c.getPath(), "animalHistory.view")).toString();
                    hrefForAnimalAbstract.append("?#subjects:").append(animalEncoded).append("&inputType:singleSubject&showReport:1&activeReport:abstract");
                    StringBuilder animalId = new StringBuilder();
                    animalId.append("<a href='").append(hrefForAnimalAbstract).append("'>");
                    animalId.append(animalEncoded);
                    animalId.append("</a>");
                    

                    msg.append(createTableBorder(behaviorsDashLength));

                    msg.append("<tr>");
                    msg.append("<td>|</td>");
                    msg.append("<td>").append(animalId).append("</td>");
                    msg.append("<td>|</td>");

Reply:
Yes, that should be safe.